bgpd/bgp_conditional_adv.c

```diff
@@ -0,0 +1,282 @@
+				return ret;
+		}
+	}
+	return ret;
+}
+
+static void bgp_conditional_adv_routes(struct peer *peer, afi_t afi,
+				       safi_t safi, struct bgp_table *table,
+				       struct route_map *rmap,
+				       enum update_type advertise)
+{
+	int addpath_capable;
+	const struct prefix *dest_p;
+	struct attr dummy_attr, attr;
+	struct bgp_path_info path;
+	struct bgp_path_info *pi;
+	struct peer_af *paf = NULL;
+	struct bgp_dest *dest = NULL;
+	struct update_subgroup *subgrp = NULL;
+
+	paf = peer_af_find(peer, afi, safi);
+	if (!paf)
+		return;
+
+	subgrp = PAF_SUBGRP(paf);
+	/* Ignore if subgroup doesn't exist (implies AF is not negotiated) */
+	if (!subgrp)
+		return;
+
+	addpath_capable = bgp_addpath_encode_tx(peer, afi, safi);
+
+	for (dest = bgp_table_top(table); dest; dest = bgp_route_next(dest)) {
+		dest_p = bgp_dest_get_prefix(dest);
+		if (!dest_p)
+			continue;
+
+		for (pi = bgp_dest_get_bgp_path_info(dest); pi; pi = pi->next) {
+			dummy_attr = *pi->attr;
+			path.peer = pi->peer;
+			path.attr = &dummy_attr;
+
+			if (route_map_apply(rmap, dest_p, RMAP_BGP, &path)
+			    != RMAP_PERMITMATCH)
+				continue;
+
+			if (CHECK_FLAG(pi->flags, BGP_PATH_SELECTED)
+			    || (addpath_capable
+				&& bgp_addpath_tx_path(
+					   peer->addpath_type[afi][safi],
+					   pi))) {
+
+				/* Skip route-map checks in
+				 * subgroup_announce_check while executing from
+				 * the conditional advertise scanner process.
+				 * otherwise when route-map is also configured
+				 * on same peer, routes in advertise-map may not
+				 * be advertised as expected.
+				 */
+				if ((advertise == ADVERTISE)
+				    && subgroup_announce_check(dest, pi, subgrp,
+							       dest_p, &attr,
+							       true))
+					bgp_adj_out_set_subgroup(dest, subgrp,
+								 &attr, pi);
+				else {
+					/* If default originate is enabled for
+					 * the peer, do not send explicit
+					 * withdraw. This will prevent deletion
+					 * of default route advertised through
+					 * default originate.
+					 */
+					if (CHECK_FLAG(
+						    peer->af_flags[afi][safi],
+						    PEER_FLAG_DEFAULT_ORIGINATE)
+					    && is_default_prefix(dest_p))
+						break;
+
+					bgp_adj_out_unset_subgroup(
+						dest, subgrp, 1,
+						bgp_addpath_id_for_peer(
+							peer, afi, safi,
+							&pi->tx_addpath));
+				}
+			}
+		}
+	}
+}
+
+/* Handler of conditional advertisement timer event.
+ * Each route in the condition-map is evaluated.
+ */
+static int bgp_conditional_adv_timer(struct thread *t)
+{
+	afi_t afi;
+	safi_t safi;
+	int pfx_rcd_safi;
+	struct bgp *bgp = NULL;
+	struct peer *peer = NULL;
+	struct peer_af *paf = NULL;
+	struct bgp_table *table = NULL;
+	struct bgp_filter *filter = NULL;
+	struct listnode *node, *nnode = NULL;
+	struct update_subgroup *subgrp = NULL;
+	route_map_result_t ret;
+
+	bgp = THREAD_ARG(t);
+	assert(bgp);
+
+	bgp->t_condition_check = NULL;
+	thread_add_timer(bm->master, bgp_conditional_adv_timer, bgp,
```

Comment:
I'm not sure why is this needed here?

Reply:
here we are creating thread (enable polling) which polls each peer (and for each AFI/SAFI within peer) in BGP process every 60 seconds once. this happens only once when the first advertise-map is created in BGP process.